src/data/int/log.lean

```diff
@@ -0,0 +1,238 @@
+
+/-!
+# Integer logarithms in a field with respect to a natural base
+
+This file defines two `ℤ`-valued analogs of the logarithm of `n : R` with base `b : ℕ`:
+
+* `int.log b n`: Lower logarithm, or floor **log**. Greatest `k` such that `↑b^k ≤ n`.
+* `int.clog b n`: Upper logarithm, or **c**eil **log**. Least `k` such that `n ≤ ↑b^k`.
+
+Note that `int.log` gives the position of the left-most non-zero digit:
+```lean
+#eval (int.log 10 (0.09 : ℚ), int.log 10 (0.10 : ℚ), int.log 10 (0.11 : ℚ))
+--    (-2,                    -1,                     -1)
+#eval (int.log 10 (9 : ℚ),    int.log 10 (10 : ℚ),   int.log 10 (11 : ℚ))
+--    (0,                      1,                     1)
+```
+which means it can be used for computing digit expansions
+```lean
+import data.fin.vec_notation
+
+def digits (b : ℕ) (q : ℚ) (n : ℕ) : ℕ :=
+⌊q*b^(↑n - int.log b q)⌋₊ % b
+
+#eval digits 10 (1/7) ∘ (coe : fin 8 → ℕ)
+-- ![1, 4, 2, 8, 5, 7, 1, 4]
+```
+
+## Main results
+
+* For `int.log`:
+  * `int.zpow_log_le_self`, `int.lt_zpow_succ_log_self`: the bounds formed by `int.log`,
+    `(b : R) ^ log b r ≤ r < (b : R) ^ (log b r + 1)`.
+  * `int.zpow_log_le_self`, `int.lt_zpow_succ_log_self`: the bounds formed by `int.log`
+  * `int.zpow_le_iff_le_log`: the galois almost-connection
+* For `int.clog`:
+  * `int.zpow_pred_clog_lt_self`, `int.self_le_zpow_clog`: the bounds formed by `int.log`,
+    `(b : R) ^ (clog b r - 1) < r ≤ (b : R) ^ clog b r`.
+  * `int.zpow_log_le_self`, `int.lt_zpow_succ_log_self`: the bounds formed by `int.log`
+  * `int.le_zpow_iff_clog_le`: the galois almost-connection
+* `int.clog_eq_neg_log_inv`: the link between the two definitions
+
+-/
+variables {R : Type*} [linear_ordered_field R] [floor_ring R]
+
+namespace int
+
+/-- The greatest power of `b` such that `b ^ log b r ≤ r`. -/
+def log (b : ℕ) (r : R) : ℤ :=
+if 1 ≤ r then
+  nat.log b (nat.floor r)
+else
+  -nat.clog b (nat.ceil r⁻¹)
+
+@[simp] lemma log_nat_cast (b : ℕ) (n : ℕ) : log b (n : R) = nat.log b n :=
+begin
+  rw log,
+  cases n,
+  { simp [nat.log_zero_right] },
+  { simp [←nat.cast_succ] }
+end
+
+lemma log_of_left_le_one {b : ℕ} (hb : b ≤ 1) (n : R) : log b n = 0 :=
+begin
+  rw log,
+  split_ifs with hn,
+  { rw [nat.log_of_left_le_one hb, int.coe_nat_zero] },
+  { rw [nat.clog_of_left_le_one hb, int.coe_nat_zero, neg_zero] },
+end
+
+lemma log_of_right_le_zero {n : R} (hn : n ≤ 0) (b : ℕ) : log b n = 0 :=
+by rw [log, if_neg (hn.trans_lt zero_lt_one).not_le,
+    nat.clog_of_right_le_one (le_of_eq_of_le (nat.ceil_eq_zero.mpr $ inv_nonpos.2 hn) zero_le_one),
+    int.coe_nat_zero, neg_zero]
+
+lemma zpow_log_le_self (b : ℕ) (r : R) (hn : 1 < b) (hr : 0 < r) :
+  (b : R) ^ log b r ≤ r :=
+begin
+  rw log,
+  split_ifs with hr1,
+  { refine le_trans _ (nat.floor_le hr.le),
+    rw [zpow_coe_nat, ←nat.cast_pow, nat.cast_le],
+    exact nat.pow_log_le_self hn (nat.floor_pos.mpr hr1) },
+  { have : 1 < r⁻¹,
+    { rw not_le at hr1,
+      exact one_lt_inv hr hr1, },
```

Comment:
```suggestion
  { have : 1 < r⁻¹ := one_lt_inv hr (not_le.1 hr1),
```
but if you were doing `obtain hr1 | hr1 := lt_or_le 1 r` instead of `split_ifs with hr1`, you wouldn't even need to use `not_le`.